### ISSUE TITLE
Add Opacus-based privacy trainers for differential privacy training

### ISSUE THREAD

**Issue by aymuos15** [aymuos15/privseg#2]
# Add Opacus Privacy Trainers for Differential Privacy Training

## Overview
This issue proposes adding differential privacy (DP) support to the PrivSeg project using [Opacus](https://opacus.ai), Meta AI's PyTorch library for training models with formal privacy guarantees. This will enable privacy-preserving training on sensitive medical imaging data while maintaining strong (ε, δ)-differential privacy guarantees.

## Background: What is Opacus?

**Opacus** is an open-source library that implements Differentially Private Stochastic Gradient Descent (DP-SGD) with minimal code changes. It provides:
- Formal mathematical privacy guarantees (epsilon, delta)
- Per-sample gradient computation and clipping
- Calibrated noise injection during training
- Real-time privacy budget tracking
- Compatibility with standard PyTorch models

### How Opacus Works

Opacus implements DP-SGD through three core mechanisms:

1. **Per-Sample Gradient Computation**: Computes gradients for each sample individually (not batched)
2. **Gradient Clipping**: Clips each gradient to maximum L2 norm (`max_grad_norm`) to bound sensitivity
3. **Noise Addition**: Adds calibrated Gaussian noise to averaged gradients proportional to `noise_multiplier`

This ensures that changing a single training sample has bounded impact on the model, providing differential privacy.

## Motivation for Medical Imaging

Medical imaging datasets contain highly sensitive patient data. Traditional deep learning training can memorize training samples, creating privacy risks. Differential privacy provides:
- **Formal Privacy Guarantees**: Mathematical proof that individual patient data is protected
- **Regulatory Compliance**: Helps meet HIPAA, GDPR requirements
- **Trust**: Enables collaboration and data sharing with privacy assurances
- **Research Value**: Quantifies privacy-utility tradeoffs in medical image segmentation

### New Training Scripts

Create privacy-enabled variants of existing training scripts:

#### 1. `src/train_private.py`
Privacy-aware U-Net training with Opacus integration:
```python
from opacus import PrivacyEngine
from opacus.validators import ModuleValidator

# Fix model for DP compatibility (replace BatchNorm with GroupNorm)
model = ModuleValidator.fix(model)

# Setup privacy engine
privacy_engine = PrivacyEngine()

# Make model, optimizer, and dataloader private
model, optimizer, train_loader = privacy_engine.make_private_with_epsilon(
    module=model,
    optimizer=optimizer,
    data_loader=train_loader,
    epochs=20,
    target_epsilon=8.0,      # Strong privacy for medical data
    target_delta=1e-5,       # < 1/dataset_size
    max_grad_norm=1.0,       # Critical hyperparameter to tune
)

# Training loop (mostly unchanged)
for epoch in range(epochs):
    for images, masks in train_loader:
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, masks)
        loss.backward()
        optimizer.step()
    
    # Track privacy budget
    epsilon = privacy_engine.get_epsilon(delta=1e-5)
    print(f"Epoch {epoch+1}: ε = {epsilon:.2f}")
```

#### 2. `src/train_mask2former_private.py`
Privacy-aware Mask2Former training with same approach

### Extended Model Comparison

Update `src/compare_models.py` to evaluate **4 model variants**:
1. U-Net (standard)
2. U-Net (private, ε=8)
3. Mask2Former (standard)
4. Mask2Former (private, ε=8)

Add privacy metrics to comparison output:
- Final epsilon (ε) and delta (δ) values
- Privacy budget consumption over training
- Privacy-utility tradeoff analysis (Dice/IoU vs epsilon)

### Required Model Fixes

**Issue**: Current U-Net uses `BatchNorm2d` (incompatible with Opacus)
- Lines 10, 13 in `src/model.py` use `nn.BatchNorm2d(out_channels)`

**Solution**: Replace with GroupNorm or use `ModuleValidator.fix()`
```python
# Option 1: Manual replacement
nn.GroupNorm(num_groups=32, num_channels=out_channels)

## Recommended Privacy Parameters

For medical imaging with strong privacy requirements:

| Parameter | Value | Rationale |
|-----------|-------|-----------|
| **Epsilon (ε)** | 8.0 | Strong privacy guarantee (1-10 is considered strong) |
| **Delta (δ)** | 1e-5 | < 1/dataset_size (for ~10k images) |
| **Max Grad Norm** | 1.0 | Start here, tune via grid search (0.1, 0.5, 1.0, 2.0, 5.0) |
| **Epochs** | 20 | Monitor epsilon; early stop if budget exhausted |
| **Batch Size** | 8 | Balance memory and privacy; larger = better gradients |
| **Learning Rate** | 1e-4 | May need adjustment from non-private training |

### Expected Performance Impact

Based on literature for medical image segmentation with DP:
- **Dice Score Impact**: 5-15% reduction with ε=8
  - Non-private: ~0.85-0.90
  - Private (ε=8): ~0.75-0.85
- **Training Time**: 1.5-3x slower due to per-sample gradient computation
- **Memory**: 1.5-2x increase; may need smaller batches or BatchMemoryManager

## Implementation Checklist

### Phase 1: Setup
- [ ] Add `opacus>=1.4.0` to `pyproject.toml` dependencies
- [ ] Install Opacus: `pip install opacus`

### Phase 2: Model Compatibility
- [ ] Fix U-Net BatchNorm issue in `src/model.py`
  - [ ] Replace `nn.BatchNorm2d` with `nn.GroupNorm` or use `ModuleValidator.fix()`
- [ ] Validate Mask2Former compatibility with Opacus
  - [ ] Test with `ModuleValidator.validate(model)`
  - [ ] Apply fixes if needed

### Phase 3: Privacy Training Scripts
- [ ] Create `src/train_private.py`
  - [ ] Import PrivacyEngine and ModuleValidator
  - [ ] Fix model with `ModuleValidator.fix()`
  - [ ] Setup privacy engine with `make_private_with_epsilon`
  - [ ] Modify training loop to track epsilon
  - [ ] Add privacy metadata to saved checkpoints
- [ ] Create `src/train_mask2former_private.py`
  - [ ] Same structure as U-Net privacy trainer
  - [ ] Handle transformer-specific considerations

### Phase 4: Hyperparameter Tuning
- [ ] Grid search over `max_grad_norm` (0.1, 0.5, 1.0, 2.0, 5.0)
- [ ] Test different epsilon values (1, 3, 5, 8, 10)
- [ ] Tune learning rate for DP training
- [ ] Experiment with batch sizes (4, 8, 16)

### Phase 5: Extended Comparison
- [ ] Update `src/compare_models.py`
  - [ ] Add support for loading private model checkpoints
  - [ ] Include privacy budget reporting (epsilon, delta)
  - [ ] Generate privacy-utility tradeoff plots
  - [ ] Compare which architecture (U-Net vs Mask2Former) is more robust to DP noise
- [ ] Create evaluation scripts for privacy analysis
  - [ ] Plot Dice/IoU vs epsilon
  - [ ] Analyze convergence curves for private training
  - [ ] Compare privacy budget consumption rates

### Phase 6: Documentation
- [ ] Update README with privacy training instructions
- [ ] Document privacy parameter selection guidelines
- [ ] Include interpretation of privacy guarantees (what ε=8 means)

## Technical Considerations

### Memory Management
Opacus requires more memory due to per-sample gradients. 

### Pre-training Strategy
To improve privacy-utility tradeoff:
1. Pre-train on public datasets (ImageNet, etc.) without DP
2. Fine-tune with DP on private medical data
3. This significantly improves final model accuracy

## References

- [Opacus Documentation](https://opacus.ai)
- [Opacus GitHub](https://github.com/pytorch/opacus)
- [DP-SGD Paper](https://arxiv.org/abs/1607.00133)
- [Medical Imaging with DP](https://blog.openmined.org)
